Sources/_StringProcessing/Compiler.swift

```diff
@@ -108,11 +108,86 @@ class Compiler {
       // We stick quoted content into read-only constant strings
       builder.buildMatchSequence(q.literal)
 
+    case .atom(let a) where a.assertionKind != nil:
+      try emitAssertion(a.assertionKind!)
+
     case .customCharacterClass, .atom:
       throw unsupported(node._dumpBase)
     }
   }
 
+  func emitAssertion(_ kind: AST.Atom.AssertionKind) throws {
+    // FIXME: Depends on API model we have... We may want to
+    // think through some of these with API interactions in mind
+    //
+    // This might break how we use `bounds` for both slicing
+    // and things like `firstIndex`, that is `firstIndex` may
+    // need to supply both a slice bounds and a per-search bounds.
+    switch kind {
+    case .startOfSubject:
+      builder.buildAssert { (input, pos, bounds) in
+        pos == input.startIndex
+      }
+
+    case .endOfSubjectBeforeNewline:
+      builder.buildAssert { (input, pos, bounds) in
+        if pos == input.endIndex { return true }
+        return input.index(after: pos) == input.endIndex
+         && input[pos].isNewline
+      }
+
+    case .endOfSubject:
+      builder.buildAssert { (input, pos, bounds) in
+        pos == input.endIndex
+      }
+
+    case .resetStartOfMatch:
+      // FIXME: Figure out how to communicate this out
+      throw unsupported(#"\K (reset/keep assertion)"#)
+
+    case .firstMatchingPositionInSubject:
+      // TODO: We can probably build a nice model with API here
+      builder.buildAssert { (input, pos, bounds) in
+        pos == bounds.lowerBound
+      }
+
+    case .textSegment:
+      // This we should be able to do!
+      throw unsupported(#"\y (text segment)"#)
```

Comment:
(please ignore until you're back from break)

@Azoy, here's where we could use some simple on-the-fly grapheme breaking. By default, `\y` will match in-between two grapheme clusters.